### PR TITLE
Mark email as not required parameter in BaseComment

### PR DESCRIPTION
**Pull request by jerry-shao**
### Tasks
* Mark email as not required parameter in BaseComment
* Added corresponding unit tests
* Added corresponding database migration scripts

### Future work
* Remove required flag for email field in frontend templates

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -326,7 +326,7 @@
                       value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF"/>
         </module>
         <module name="JavadocMethod">
-            <property name="scope" value="public"/>
+            <property name="accessModifiers" value="public"/>
```
Comment (ruibaby) [inline config/checkstyle/checkstyle.xml:329]: this modification is not related to the current pr

File: src/main/java/run/halo/app/model/properties/CommentProperties.java [hunk 2]
```diff
@@ -14,6 +14,11 @@
      */
     GRAVATAR_DEFAULT("comment_gravatar_default", String.class, ""),
 
+    /**
+     * The identicon gravatar type.
+     */
+    GRAVATAR_IDENTICON("comment_gravatar_identicon", String.class, "identicon"),
```
Comment (ruibaby) [inline src/main/java/run/halo/app/model/properties/CommentProperties.java:20]: I don’t think it’s necessary, just use `GRAVATAR_DEFAULT` directly

**Review by ruibaby — CHANGES_REQUESTED** [review]

File: src/main/resources/migration/V5__migrate_1.4.13_to_1.4.14.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+-- Migrate 1.4.13 to 1.4.14
```
Comment (ruibaby) [inline src/main/resources/migration/V5__migrate_1.4.13_to_1.4.14.sql:1]: Naming recommendations for migration files are named according to what has been specifically modified, not versions

**Review by ruibaby — CHANGES_REQUESTED** [review]

File: src/main/java/run/halo/app/model/properties/CommentProperties.java (hunk 2 above)
Comment (jerry-shao) [reply src/main/java/run/halo/app/model/properties/CommentProperties.java]: Fixed.

File: config/checkstyle/checkstyle.xml (hunk 1 above)
Comment (jerry-shao) [reply config/checkstyle/checkstyle.xml]: Fixed.

File: src/main/resources/migration/V5__migrate_1.4.13_to_1.4.14.sql (hunk 3 above)
Comment (jerry-shao) [reply src/main/resources/migration/V5__migrate_1.4.13_to_1.4.14.sql]: Updated the filename and comment in the file.

**Review by ruibaby — APPROVED** [review]

File: src/main/resources/migration/V5__migrate_remove_notnull_for_email_in_comments_table.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+-- Remove notnull for email in comments table
+
+-- Migrate comments Table
+alter table comments modify email VARCHAR(255);
```
Comment (JohnNiang) [inline src/main/resources/migration/V5__migrate_remove_notnull_for_email_in_comments_table.sql:4]: Newline needed.

**Review by JohnNiang — CHANGES_REQUESTED** [review]
Hi @jerry-shao , I suggest testing against comment API without providing email.

File: src/main/resources/migration/V5__migrate_remove_notnull_for_email_in_comments_table.sql (hunk 4 above)
Comment (jerry-shao) [reply src/main/resources/migration/V5__migrate_remove_notnull_for_email_in_comments_table.sql:4]: Fixed

**Comment by jerry-shao** [comment]
> Hi @jerry-shao , I suggest testing against comment API without providing email.

Good catch, a NPE got captured by adding more tests.

**Review by guqing — APPROVED** [review]

File: src/test/java/run/halo/app/model/params/validation/PostCommentParamValidationTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private PostCommentParam postCommentParam;
+
+    @BeforeEach
+    public void setUp() {
+        validator = factory.getValidator();
+        postCommentParam = new PostCommentParam();
+    }
+
+    @Test
+    public void validate_whenEmailIsNull_NoValidationViolations() {
```
[the remark this answers is not in the document]
Comment (jerry-shao) [reply src/test/java/run/halo/app/model/params/validation/PostCommentParamValidationTest.java]: Fixed.

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        MockitoAnnotations.openMocks(this);
+        postController = new PostController(
+            mockPostService,
+            mockPostCommentService,
+            mockOptionService
+        );
+    }
+
+    @Test
+    public void nullEmailCommentTest() {
```
Comment (JohnNiang) [inline src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:53]: Could you please provide more test cases here to guarantee that previous comment logic works well?

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    @Mock
+    private OptionService mockOptionService;
+
+    @Mock
+    private BaseCommentDTO mockBaseCommentDTO;
+
+    @Mock
+    private PostComment mockPostComment;
+
+    private PostController postController;
```
Comment (JohnNiang) [inline src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:40]: Meanwhile, we are not recommended that invoke controller method directly to test controller.

In this way, the parameters of API won't be checked really.

Please see also https://github.com/halo-dev/halo/blob/master/src/test/java/run/halo/app/it/BaseApiTest.java

**Review by JohnNiang — COMMENTED** [review]
Hi @jerry-shao , I've left some comments below, please take a moment to look at them.

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java (hunk 7 above)
Comment (JohnNiang) [reply src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:40]: New references: https://spring.io/guides/gs/testing-web/

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java (hunk 6 above)
Comment (jerry-shao) [reply src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:71]: Fixed in the new revision.

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java (hunk 7 above)
Comment (jerry-shao) [reply src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:42]: Added corresponding integration tests for comment API.

**Comment by JohnNiang** [comment]
Thank @jerry-shao  for your contribution! I'll look at the new IT test tomorrow carefully.

BTW, please fix the check style conflicts first.

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+import run.halo.app.model.dto.BaseCommentDTO;
+import run.halo.app.model.entity.PostComment;
+import run.halo.app.model.params.PostCommentParam;
+import run.halo.app.service.OptionService;
+import run.halo.app.service.PostCommentService;
+import run.halo.app.service.PostService;
+
+@SpringBootTest
+@ActiveProfiles("test")
+public class PostControllerTest {
```
Comment (JohnNiang) [inline src/test/java/run/halo/app/controller/content/api/PostControllerTest.java:21]: Hi @jerry-shao , could you remove this `PostControllerTest`? Because you have provided `PostAPITest` already.

**Review by JohnNiang — CHANGES_REQUESTED** [review]
Hi @jerry-shao , I've left one comment here. Everything is ok to me but this.

File: src/test/java/run/halo/app/controller/content/api/PostControllerTest.java (hunk 8 above)
Comment (jerry-shao) [reply src/test/java/run/halo/app/controller/content/api/PostControllerTest.java]: Removed.

**Comment by JohnNiang** [comment]
LGTM

**Review by JohnNiang — APPROVED** [review]
